modules/apps/object/object-web/src/main/java/com/liferay/object/web/internal/object/entries/upload/AttachmentUploadFileEntryHandler.java

```diff
@@ -0,0 +1,116 @@
+/**
+ * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
+ *
+ * This library is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU Lesser General Public License as published by the Free
+ * Software Foundation; either version 2.1 of the License, or (at your option)
+ * any later version.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+package com.liferay.object.web.internal.object.entries.upload;
+
+import com.liferay.object.model.ObjectDefinition;
+import com.liferay.object.model.ObjectField;
+import com.liferay.object.service.ObjectDefinitionLocalService;
+import com.liferay.object.service.ObjectFieldLocalService;
+import com.liferay.object.web.internal.object.entries.upload.util.AttachmentUploadValidator;
+import com.liferay.portal.kernel.exception.PortalException;
+import com.liferay.portal.kernel.portletfilerepository.PortletFileRepository;
+import com.liferay.portal.kernel.repository.model.FileEntry;
+import com.liferay.portal.kernel.theme.ThemeDisplay;
+import com.liferay.portal.kernel.upload.UploadPortletRequest;
+import com.liferay.portal.kernel.util.FileUtil;
+import com.liferay.portal.kernel.util.MimeTypes;
+import com.liferay.portal.kernel.util.ParamUtil;
+import com.liferay.portal.kernel.util.WebKeys;
+import com.liferay.upload.UploadFileEntryHandler;
+
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Reference;
+
+/**
+ * @author Carolina Barbosa
+ */
+@Component(service = AttachmentUploadFileEntryHandler.class)
+public class AttachmentUploadFileEntryHandler
+	implements UploadFileEntryHandler {
+
+	@Override
+	public FileEntry upload(UploadPortletRequest uploadPortletRequest)
+		throws IOException, PortalException {
+
+		File file = null;
+
+		try (InputStream inputStream = uploadPortletRequest.getFileAsStream(
+				"file")) {
+
+			file = FileUtil.createTempFile(inputStream);
+
+			String fileName = uploadPortletRequest.getFileName("file");
+
+			long objectFieldId = ParamUtil.getLong(
+				uploadPortletRequest, "objectFieldId");
+
+			_attachmentUploadValidator.validateFileSize(
+				file, fileName, objectFieldId);
+
+			_attachmentUploadValidator.validateFileExtension(
+				fileName, objectFieldId);
+
+			long folderId = ParamUtil.getLong(uploadPortletRequest, "folderId");
```

Comment:
Should we validate if the folder is the correct folder assigned to that Object? I mean, malicious users can tamper the upload request and change the destination.

Reply:
I did some tests and the `_portletFileRepository.addPortletFileEntry` already does this validation (throwing a NoSuchFolderException), because I'm passing the `objectDefinition.getPortletId()`. 
But I had to add a `ModelResourcePermissionUtil.check`.
Thanks for bringing this to my attention 😄

Reply:
Nice! As I was answering the other comment the url, I was thinking that maybe we should think about getting the folderId and objectFieldId from the Session and not the request. What do you think?

Reply:
As offline discussed, let's not consider this issues for now.